### ISSUE TITLE
Fix Doxygen coverage

### ISSUE THREAD

**Issue by jmpf** [Chaste/Chaste#179]
See #160 for details of automatic generation of Doxygen pages.

From @fcooper8472 :
"You can see the failures on the actions runs, e.g. here, https://github.com/Chaste/Chaste/actions/runs/6383539495/job/17324354869?pr=158, and there are a large number of failures that I don't understand."

**Comment by jmpf** [comment]
Now working on this in branch [issue_179_doxygen](https://github.com/Chaste/Chaste/tree/issue_179_doxygen)

**Comment by fcooper8472** [comment]
There was already a branch for doxygen:
https://github.com/Chaste/Chaste/tree/doxygen

I've made several commits that address some of the cell based and global doxygen coverage errors. Just want to make sure we're not duplicating too much effort.

https://github.com/Chaste/Chaste/pull/158/commits/a8adb82317e36f5599b9c6b13fe53f929b1b6d39
https://github.com/Chaste/Chaste/pull/158/commits/9303c0bc92bcf637773f236c263ae6dfef6ff89d

**Comment by jmpf** [comment]
I saw that.  I thought that it might be best to separate infrastructure (your branch) from tidying (my branch).  It may be slicker for testing if I merge/delete my stuff though...

**Comment by jmpf** [comment]
Sorry.  I see.  I have duplicated effort.  In my defence none of your work on fixing coverage is documented in either issue.

**Comment by fcooper8472** [comment]
> Sorry. I see. I have duplicated effort. In my defence none of your work on fixing coverage is documented in either issue.

Entirely my fault for not adequately documenting what I was doing!

**Comment by fcooper8472** [comment]
@jmpf see, e.g., this result:
https://github.com/Chaste/Chaste/actions/runs/6433492539/job/17470665049

If you look at the "print doxygen coverage errors" step, you'll see near the top:
"Contents of doxygen/coverage/cell_based-src-population-AbstractCellPopulation.cpp.52.0:"

The 52 here indicates 52 errors, which is as far as I can tell, every method in the file. It somehow can't uniquely match the definitions to the declarations in the header, but I can't see any reason why not!

We're using the following doxygen:

```
wget https://github.com/doxygen/doxygen/releases/download/Release_1_9_8/doxygen-1.9.8.linux.bin.tar.gz
tar -xzvf doxygen-1.9.8.linux.bin.tar.gz
```

**Comment by jmpf** [comment]
Reproduced locally.  Number of errors and output of errors is different between Doxygen versions.  I have switched from Doxygen 1.9.1-2ubuntu2 to doxygen-1.9.8:
`~/doxygen-1.9.8/bin/doxygen Doxyfile 2>err 1>out && python python/infra/ParseDoxygen.py out err coverage`

**Comment by jmpf** [comment]
It's not https://github.com/doxygen/doxygen/issues/8497 (which is an unresolved issue to do with overloading methods).

**Comment by fcooper8472** [comment]
Progress, but no fix yet:

- The problem exists with all the versions of Doxygen I have tried dating back several years.
- The problem exists when using Doxygen with `CLANG_ASSISTED_PARSING` set to "YES"; so it's likely NOT a problem with how doxygen is parsing the files

The main stride forward in understand this is that many of the errors exist when there is a default template parameter, e.g. of this form:

```
template<unsigned ELEMENT_DIM, unsigned SPACE_DIM = ELEMENT_DIM>
```

Removing the default value (the `= ELEMENT_DIM` part) resolves the problems.

Interestingly, adding the `= ELEMENT_DIM` to the definitions in the .cpp files does NOT resolve the problem.

**Comment by mirams** [comment]
update needed here @fcooper8472 from what you were saying in the meeting?

**Comment by fcooper8472** [comment]
This is done: fixed by the commits referenced in #199
